src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt

```diff
@@ -0,0 +1,567 @@
+            )
+        }
+
+        private fun errorText(): String {
+            val lifetimeText = escapeString(element.text)
+            return "Use of undeclared lifetime name `$lifetimeText`"
+        }
+    }
+}
+
+sealed class RsTextRangeDiagnostic(val element: TextRange) {
+    abstract fun addToHolder(holder: AnnotationHolder)
+
+    class TraitItemsMissingImplError(
+        element: TextRange,
+        private val missing: String,
+        private val impl: RsImplItem
+    ) : RsTextRangeDiagnostic(element) {
+        override fun addToHolder(holder: AnnotationHolder) {
+            holder.createErrorAnnotation(
+                element,
+                header(E0046, errorText())
+            ).registerFix(ImplementMembersFix(impl))
+        }
+
+        private fun errorText(): String {
+            val missingS = escapeString(missing)
+            return "Not all trait items implemented, missing: $missingS"
+        }
+    }
+
+    class CrateNotFoundError(
+        element: TextRange,
+        private val crateName: String
+    ) : RsTextRangeDiagnostic(element) {
+        override fun addToHolder(holder: AnnotationHolder) {
+            holder.createErrorAnnotation(
+                element,
+                header(E0463, errorText())
+            )
+        }
+
+        private fun errorText(): String {
+            val elTextS = escapeString(crateName)
+            return "Can't find crate for `$elTextS`"
+        }
+    }
+}
+
+private class AnnotationHeader(
+    val plainText: String,
+    val html: String
+)
+
+enum class RsErrorCode {
+    E0046, E0050, E0060, E0061, E0069,
+    E0121, E0124, E0133, E0185, E0186, E0198, E0199,
+    E0200, E0201, E0202, E0261, E0263,
+    E0308, E0379,
+    E0403, E0407, E0415, E0424, E0426, E0428, E0449, E0463,
+    E0603, E0616, E0624;
+
+    val code: String
+        get() = toString()
+    val infoUrl: String
+        get() = "https://doc.rust-lang.org/error-index.html#$code"
+}
+
+private fun header(error: RsErrorCode, description: String): AnnotationHeader =
+    AnnotationHeader("$description ${error.code}", "$description <a href='${error.infoUrl}'>${error.code}</a>")
+
+private fun AnnotationHolder.createErrorAnnotation(
+    element: PsiElement,
+    header: AnnotationHeader,
+    detailsHtml: String = ""
+): Annotation {
+    return this.createErrorAnnotation(
+        element.textRange,
+        header,
+        detailsHtml
+    )
+}
+
+private fun AnnotationHolder.createErrorAnnotation(
+    textRange: TextRange,
+    header: AnnotationHeader,
+    detailsHtml: String = ""
+): Annotation {
+    return createAnnotation(HighlightSeverity.ERROR,
+        textRange,
+        header.plainText,
+        "<html>${header.html}<br>$detailsHtml</html>"
+    )
+}
+
+private fun AnnotationHolder.requireResolve(el: RsReferenceElement, header: AnnotationHeader, message: String = "") {
+    if (el.reference.resolve() != null || el.reference.multiResolve().size > 1) return
```

Comment:
Hmm, there is still some inconsistency. We creating `RsDiagnostics` first, then converting it to `Annotation`. But here we may or **may not** create an annotation due to this condition. Can you please try to move it on the upper level, before creating `RsDiagnostics`?

Reply:
Sorry, I must have missed the notification for this. I will make this change shortly.

Reply:
Oh, and now it need to be rebased... There was an API changes